src/main/java/com/iota/iri/hash/PearlDiver.java

```diff
@@ -44,179 +81,135 @@ public synchronized boolean search(final int[] transactionTrits, final int minWe
             throw new RuntimeException("Invalid min weight magnitude: " + minWeightMagnitude);
         }
 
+        // state still needs protected, because it is 
+        // possible that another thread calls cancel()
         synchronized (syncObj) {
             state = RUNNING;
         }
 
         final long[] midCurlStateLow = new long[CURL_STATE_LENGTH], midCurlStateHigh = new long[CURL_STATE_LENGTH];
+        int numberOfThreads = Math.max(Runtime.getRuntime().availableProcessors() - 1, 1);
+        Thread[] workers = new Thread[numberOfThreads];
 
-        {
-            for (int i = CURL_HASH_LENGTH; i < CURL_STATE_LENGTH; i++) {
-                midCurlStateLow[i] = HIGH_BITS;
-                midCurlStateHigh[i] = HIGH_BITS;
-            }
-
-            int offset = 0;
-            final long[] curlScratchpadLow = new long[CURL_STATE_LENGTH], curlScratchpadHigh = new long[CURL_STATE_LENGTH];
-            for (int i = (TRANSACTION_LENGTH - CURL_HASH_LENGTH) / CURL_HASH_LENGTH; i-- > 0; ) {
-
-                for (int j = 0; j < CURL_HASH_LENGTH; j++) {
-
-                    switch (transactionTrits[offset++]) {
-                        case 0: {
-                            midCurlStateLow[j] = HIGH_BITS;
-                            midCurlStateHigh[j] = HIGH_BITS;
-
-                        }
-                        break;
+        para(transactionTrits, midCurlStateLow, midCurlStateHigh);
 
-                        case 1: {
-                            midCurlStateLow[j] = LOW_BITS;
-                            midCurlStateHigh[j] = HIGH_BITS;
-                        }
-                        break;
+        while (numberOfThreads-- > 0) {
+            final int threadIndex = numberOfThreads;
+            final long[] midCurlStateCopyLow = new long[CURL_STATE_LENGTH];
+            final long[] midCurlStateCopyHigh = new long[CURL_STATE_LENGTH];
 
-                        default: {
-                            midCurlStateLow[j] = HIGH_BITS;
-                            midCurlStateHigh[j] = LOW_BITS;
-                        }
-                    }
-                }
+            System.arraycopy(midCurlStateLow, 0, midCurlStateCopyLow, 0, CURL_STATE_LENGTH);
+            System.arraycopy(midCurlStateHigh, 0, midCurlStateCopyHigh, 0, CURL_STATE_LENGTH);
 
-                transform(midCurlStateLow, midCurlStateHigh, curlScratchpadLow, curlScratchpadHigh);
+            for (int i = threadIndex; i-- > 0; ) {
+                increment(midCurlStateCopyLow, midCurlStateCopyHigh, 162 + CURL_HASH_LENGTH / 9,
```

Comment:
This will mutate in the wrong location. You should only mutate the final 81 trits.